### PR TITLE
[ONNX] handle aten::_set_item on Dict in convertInplaceOpsAndTrackAlias

### PR DESCRIPTION
**Pull request by garymm**
It seems the JIT produces an output for aten::_set_item on lists but
not on dicts. Previously the code would crash because it assumed it
was operating on a list.
    
The different behavior can be seen with the following test:

```python
class DictModule(torch.nn.Module):
    def forward(self, x_in: torch.Tensor) -> typing.Dict[str, torch.Tensor]:
        x_out = {}
        x_out["test_key_out"] = x_in
        return x_out
    
x_in = torch.tensor(1)
dms = torch.jit.script(DictModule())
torch.onnx.export(dms, (x_in,), "/dev/null", example_outputs=(dms(x_in),))
```

Before this change:
`RuntimeError: outputs_.size() == 1INTERNAL ASSERT FAILED at "../torch/csrc/jit/ir/ir.h":452, please report a bug to PyTorch.`
    
After this change:
`RuntimeError: Exporting the operator prim_DictConstruct to ONNX opset version 9 is not supported. Please feel free to request support or submit a pull request on PyTorch GitHub.`
    
This is a more useful error message.

### CODE REVIEW

**Comment by BowenBao** [comment]
Thanks for the fix! PR looks good, could you include the test case in the description in unit test as well, with `assertRaises`?

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp [hunk 1]
```diff
@@ -270,8 +270,12 @@ static std::pair<Value*, Value*> PrepareListAppendAndInsertForONNX(Node* n) {
   return std::make_pair(n->input(0), n->output());
 }
 
-static std::pair<Value*, Value*> PrepareListSetItemForONNX(Node* n) {
+static std::pair<Value*, Value*> PrepareSetItemForONNX(Node* n) {
   TORCH_INTERNAL_ASSERT(n->kind() == aten::_set_item);
+  if (n->outputs().size() == 0) {
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:275]: Perhaps comment on the List vs Dict discrepancy here as to why this is needed.

Comment (garymm) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:275]: Done.

**Comment by garymm** [comment]
> Thanks for the fix! PR looks good, could you include the test case in the description in unit test as well, with `assertRaises`?

Good idea. Done!
BTW, Is there a way that we can more easily differentiate between assertion failures in the JIT and unsupported operations? Perhaps with a different exception type?

**Comment by BowenBao** [comment]
@garymm That would be a great idea, i think currently we don't have a specialized exception type for onnx export. It'd be interesting to see how to use a same exception type across C++ and Python code, maybe need to touch `torch/csrc/Exceptions.h` and make sure it works with pybind. 
Another point I see in this example is that there are cases of assertion errors that were actually caused by unsupported operations, yet it failed early at assertions before reaching the place to print a more meaningful error. I guess we need to be more careful about what we assume when writing passes.

**Comment by garymm** [comment]
Filed https://msdata.visualstudio.com/Vienna/_workitems/edit/1185416/.

Please take another look at this PR and approve and merge if ready.

Thanks!

**Pull request merged**